File: production/frontend/src/stores/bank.ts
/**
 * stores/bank.ts — 银行信任培育期渐进解锁状态管理 (CORE-01b)
 *
 * 职责:
 *   1. 维护当前选中银行 ID (持久化)
 *   2. 缓存银行列表 / 信任档案 / 风险提示函 / 决策日志 / 统计
 *   3. 暴露 8 个 API action 与后端 /bank/* 端点对齐
 *   4. 切换银行时自动清空缓存 (避免跨银行污染)
 *
 * project_memory 硬约束:
 *   - L4 培育期 AI 零拦截, 仅生成风险提示函
 *   - L2 信任期 <50 万自动放行 (50 万元 = 50,000,000 分)
 *   - 切银行时清空风险提示函/决策/统计缓存
 */

import { defineStore } from 'pinia';
import { ref } from 'vue';
import * as bankApi from '@/api/bank';
import type {
  BankDecision, BankDecisionCreate, BankListItem, BankStatistics,
  BankTrustProfile, BankTrustStage, RiskLetter, RiskLetterCreate,
  StageUpgradeResult, SupervisionAccount, CreditMultiplierPayload,
  CreditMultiplierResult, FreezeAccountPayload, FreezeAccountResult,
} from '@/api/bank';

// === 阶段元数据 (供 UI 进度条/高亮使用) ===
export const STAGE_ORDER: BankTrustStage[] = [
  'L4_READONLY', 'L3_ADVISORY', 'L2_SMALL_AUTO', 'L1_FULL_AUTO',
];

export const STAGE_LABELS: Record<BankTrustStage, string> = {
  L4_READONLY: '培育期 (0-6 月)',
  L3_ADVISORY: '验证期 (6-12 月)',
  L2_SMALL_AUTO: '信任期 (12-24 月)',
  L1_FULL_AUTO: '深度信任期 (24+ 月)',
};

export const STAGE_SHORT_LABELS: Record<BankTrustStage, string> = {
  L4_READONLY: 'L4 只读',
  L3_ADVISORY: 'L3 建议',
  L2_SMALL_AUTO: 'L2 小额自动',
  L1_FULL_AUTO: 'L1 全自动',
};

// 50 万元 = 50,000,000 分 (与后端阈值对齐)
export const SMALL_AUTO_THRESHOLD_CENTS = 50_000_000;

export const useBankStore = defineStore('bank', () => {
  // === state ===
  const currentBankId = ref<string | null>(null);
  const banks = ref<BankListItem[]>([]);
  const trustProfile = ref<BankTrustProfile | null>(null);
  const riskLetters = ref<RiskLetter[]>([]);
  const decisions = ref<BankDecision[]>([]);
  const statistics = ref<BankStatistics | null>(null);
  const loading = ref(false);
  const error = ref<string | null>(null);
  /** 风险提示函筛选企业 ID */
  const letterEnterpriseFilter = ref<string | undefined>(undefined);
  /** APP-01: 监管账户列表 */
  const supervisionAccounts = ref<SupervisionAccount[]>([]);
  /** APP-01: 当前授信乘数 (默认 1.0, 后端可调整) */
  const currentMultiplier = ref<number>(1.0);
  /** APP-01: 当前授信额度 (分) */
  const currentCreditLimitCents = ref<number>(0);

  // === helpers ===

  /** 切换银行时清空业务缓存 (保留 banks 列表) */
  function _clearBusinessCache(): void {
    trustProfile.value = null;
    riskLetters.value = [];
    decisions.value = [];
    statistics.value = null;
    letterEnterpriseFilter.value = undefined;
    supervisionAccounts.value = [];
    currentMultiplier.value = 1.0;
    currentCreditLimitCents.value = 0;
  }

  // === actions (与 8 个端点对齐) ===

  /** 端点 1: 列出所有银行 (并自动选中第一个) */
  async function fetchBanks(): Promise<BankListItem[]> {
    loading.value = true;
    error.value = null;
    try {
      const result = await bankApi.listBanks();
      banks.value = result;
      if (!currentBankId.value && result.length > 0) {
        currentBankId.value = result[0]!.bankId;
      }
      return result;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 拉取银行列表失败:', e);
      return [];
    } finally {
      loading.value = false;
    }
  }

  /** 切换当前银行 (清空业务缓存) */
  async function switchBank(bankId: string): Promise<void> {
    if (bankId === currentBankId.value) return;
    currentBankId.value = bankId;
    _clearBusinessCache();
  }

  /** 端点 2: 银行信任档案 */
  async function fetchTrustProfile(bankId?: string): Promise<BankTrustProfile | null> {
    const id = bankId ?? currentBankId.value;
    if (!id) return null;
    try {
      const profile = await bankApi.getTrustProfile(id);
      trustProfile.value = profile;
      return profile;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 拉取信任档案失败:', e);
      return null;
    }
  }

  /** 端点 3: 升级信任阶段 */
  async function upgradeStage(bankId?: string): Promise<StageUpgradeResult | null> {
    const id = bankId ?? currentBankId.value;
    if (!id) return null;
    try {
      const result = await bankApi.upgradeStage(id);
      // 升级成功后刷新信任档案
      if (result.upgraded) {
        await fetchTrustProfile(id);
        await fetchBanks();
      }
      return result;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 升级阶段失败:', e);
      return null;
    }
  }

  /** 端点 4: 风险提示函列表 (可按企业筛选) */
  async function fetchRiskLetters(
    bankId?: string,
    enterpriseId?: string,
  ): Promise<RiskLetter[]> {
    const id = bankId ?? currentBankId.value;
    if (!id) return [];
    try {
      const filterEnt = enterpriseId ?? letterEnterpriseFilter.value;
      const result = await bankApi.listRiskLetters(id, filterEnt);
      riskLetters.value = result;
      return result;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 拉取风险提示函失败:', e);
      return [];
    }
  }

  /** 设置风险提示函筛选企业 */
  function setLetterEnterpriseFilter(enterpriseId?: string): void {
    letterEnterpriseFilter.value = enterpriseId;
  }

  /** 端点 5: 生成风险提示函 (L4 培育期 AI 唯一输出) */
  async function generateRiskLetter(
    payload: RiskLetterCreate,
    bankId?: string,
  ): Promise<RiskLetter | null> {
    const id = bankId ?? currentBankId.value;
    if (!id) return null;
    try {
      const letter = await bankApi.generateRiskLetter(id, payload);
      // 写入缓存列表头部
      riskLetters.value = [letter, ...riskLetters.value];
      return letter;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 生成风险提示函失败:', e);
      return null;
    }
  }

  /** 端点 6: 决策日志 */
  async function fetchDecisions(bankId?: string): Promise<BankDecision[]> {
    const id = bankId ?? currentBankId.value;
    if (!id) return [];
    try {
      const result = await bankApi.listDecisions(id);
      decisions.value = result;
      return result;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 拉取决策日志失败:', e);
      return [];
    }
  }

  /** 端点 7: 提交决策 (根据 stage 自动判定 autoHandled) */
  async function submitDecision(
    payload: BankDecisionCreate,
    bankId?: string,
  ): Promise<BankDecision | null> {
    const id = bankId ?? currentBankId.value;
    if (!id) return null;
    try {
      const decision = await bankApi.submitDecision(id, payload);
      decisions.value = [decision, ...decisions.value];
      return decision;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 提交决策失败:', e);
      return null;
    }
  }

  /** 端点 8: 统计数据 */
  async function fetchStatistics(bankId?: string): Promise<BankStatistics | null> {
    const id = bankId ?? currentBankId.value;
    if (!id) return null;
    try {
      const stats = await bankApi.getStatistics(id);
      statistics.value = stats;
      return stats;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 拉取统计失败:', e);
      return null;
    }
  }

  /** 一键加载当前银行的全部数据 (切换银行后调用) */
  async function loadAll(bankId?: string): Promise<void> {
    const id = bankId ?? currentBankId.value;
    if (!id) return;
    await Promise.all([
      fetchTrustProfile(id),
      fetchRiskLetters(id),
      fetchDecisions(id),
      fetchStatistics(id),
      fetchSupervisionAccounts(id),
    ]);
  }

  /** APP-01 端点 9: 调整授信乘数 */
  async function adjustCreditMultiplier(
    payload: CreditMultiplierPayload,
    bankId?: string,
  ): Promise<CreditMultiplierResult | null> {
    const id = bankId ?? currentBankId.value;
    if (!id) return null;
    try {
      const result = await bankApi.adjustCreditMultiplier(id, payload);
      currentMultiplier.value = result.multiplier;
      currentCreditLimitCents.value = result.newCreditLimitCents;
      return result;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 调整授信乘数失败:', e);
      return null;
    }
  }

  /** APP-01 端点 10: 冻结/解冻监管账户 */
  async function freezeAccount(
    accountId: string,
    payload: FreezeAccountPayload,
    bankId?: string,
  ): Promise<FreezeAccountResult | null> {
    const id = bankId ?? currentBankId.value;
    if (!id) return null;
    try {
      const result = await bankApi.freezeAccount(id, accountId, payload);
      // 同步更新本地缓存
      const idx = supervisionAccounts.value.findIndex((a) => a.accountId === accountId);
      if (idx >= 0) {
        supervisionAccounts.value[idx] = {
          ...supervisionAccounts.value[idx]!,
          status: result.status,
          lastOperationAt: result.operatedAt,
        };
      }
      return result;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 冻结/解冻账户失败:', e);
      return null;
    }
  }

  /** APP-01 端点 11: 列出监管账户 */
  async function fetchSupervisionAccounts(
    bankId?: string,
  ): Promise<SupervisionAccount[]> {
    const id = bankId ?? currentBankId.value;
    if (!id) return [];
    try {
      const result = await bankApi.listSupervisionAccounts(id);
      supervisionAccounts.value = result;
      // 推断当前授信乘数 (若已有账户, 取最大授信额度对应企业, 否则默认 1.0)
      const maxBalance = result.reduce((m, a) => Math.max(m, a.balanceCents), 0);
      if (maxBalance > 0) {
        currentCreditLimitCents.value = maxBalance;
      }
      return result;
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e);
      console.error('[BankStore] 拉取监管账户失败:', e);
      return [];
    }
  }

  return {
    // state
    currentBankId,
    banks,
    trustProfile,
    riskLetters,
    decisions,
    statistics,
    loading,
    error,
    letterEnterpriseFilter,
    supervisionAccounts,
    currentMultiplier,
    currentCreditLimitCents,
    // actions
    fetchBanks,
    switchBank,
    fetchTrustProfile,
    upgradeStage,
    fetchRiskLetters,
    setLetterEnterpriseFilter,
    generateRiskLetter,
    fetchDecisions,
    submitDecision,
    fetchStatistics,
    loadAll,
    adjustCreditMultiplier,
    freezeAccount,
    fetchSupervisionAccounts,
  };
}, {
  persist: {
    key: 'fintrust-bank',
    storage: localStorage,
    paths: ['currentBankId'],
  },
});

File: production/backend/app/api/v1/bank.py
"""银行信任培育期渐进解锁 API 路由 (CORE-01b).

端点 (prefix=/bank):
    GET    /bank                                 列出所有银行
    GET    /bank/{bank_id}/trust-profile          银行信任档案
    POST   /bank/{bank_id}/upgrade-stage          升级信任阶段
    GET    /bank/{bank_id}/risk-letters           风险提示函列表 (可选 ?enterprise_id=)
    POST   /bank/{bank_id}/risk-letters           生成风险提示函
    GET    /bank/{bank_id}/decisions              决策日志
    POST   /bank/{bank_id}/decisions              提交决策
    GET    /bank/{bank_id}/statistics             统计数据
    GET    /bank/{bank_id}/supervision-accounts   监管账户列表 (APP-01 端点 11)
    POST   /bank/{bank_id}/freeze-account         冻结/解冻监管账户 (APP-01 端点 10)
    POST   /bank/{bank_id}/credit-multiplier      调整授信乘数 (APP-01 端点 9)

响应全部用 make_ok 包装, response_model 使用 ApiResult[T].
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query, status

from app.api.deps import make_ok
from app.deps import CurrentUser
from app.schemas.bank import (
    BankDecision,
    BankDecisionCreate,
    BankListItem,
    BankStatistics,
    BankTrustProfile,
    CreditMultiplierPayload,
    CreditMultiplierResult,
    FreezeAccountPayload,
    FreezeAccountResult,
    RiskLetter,
    RiskLetterCreate,
    StageUpgradeResult,
    SupervisionAccount,
)
from app.schemas.common import ApiResult
from app.services.bank_service import BankService

bank_router = APIRouter(prefix="/bank", tags=["银行信任培育 (CORE-01b)"])


def _svc() -> BankService:
    return BankService(db=None)


# ============================================================================
# 1. 列出所有银行
# ============================================================================

@bank_router.get("", response_model=ApiResult[list[BankListItem]], summary="列出所有银行")
async def list_banks(_user: CurrentUser):
    """列出所有银行 (含信任阶段, 供银行切换选择)."""
    items = await _svc().list_banks()
    return make_ok(items)


# ============================================================================
# 2. 银行信任档案
# ============================================================================

@bank_router.get(
    "/{bank_id}/trust-profile",
    response_model=ApiResult[BankTrustProfile],
    summary="银行信任档案",
)
async def get_trust_profile(bank_id: str, _user: CurrentUser):
    """获取银行信任档案 (4 阶段 + 转化率 + 解锁进度)."""
    profile = await _svc().get_trust_profile(bank_id)
    if not profile:
        return make_ok(None, code=404, message=f"银行 {bank_id} 不存在")
    return make_ok(profile)


# ============================================================================
# 3. 升级信任阶段
# ============================================================================

@bank_router.post(
    "/{bank_id}/upgrade-stage",
    response_model=ApiResult[StageUpgradeResult],
    summary="升级信任阶段",
)
async def upgrade_stage(bank_id: str, _user: CurrentUser):
    """升级信任阶段 (需满足接入月数/决策数/转化率三阈值)."""
    result = await _svc().upgrade_stage(bank_id)
    return make_ok(result)


# ============================================================================
# 4. 风险提示函列表
# ============================================================================

@bank_router.get(
    "/{bank_id}/risk-letters",
    response_model=ApiResult[list[RiskLetter]],
    summary="风险提示函列表",
)
async def list_risk_letters(
    bank_id: str,
    enterprise_id: str | None = Query(default=None, alias="enterpriseId", description="按企业筛选"),
    _user: CurrentUser = None,
):
    """风险提示函列表 (L4 培育期 AI 唯一输出, 可按企业筛选)."""
    items = await _svc().list_risk_letters(bank_id, enterprise_id)
    return make_ok(items)


# ============================================================================
# 5. 生成风险提示函
# ============================================================================

@bank_router.post(
    "/{bank_id}/risk-letters",
    response_model=ApiResult[RiskLetter],
    status_code=status.HTTP_201_CREATED,
    summary="生成风险提示函",
)
async def generate_risk_letter(
    bank_id: str, payload: RiskLetterCreate, _user: CurrentUser,
):
    """生成风险提示函 (L4 培育期零拦截)."""
    letter = await _svc().generate_risk_letter(
        bank_id=bank_id,
        enterprise_id=payload.enterprise_id,
        risk_level=payload.risk_level,
        summary=payload.summary,
        recommendations=payload.recommendations,
        enterprise_name=payload.enterprise_name,
    )
    return make_ok(letter)


# ============================================================================
# 6. 决策日志
# ============================================================================

@bank_router.get(
    "/{bank_id}/decisions",
    response_model=ApiResult[list[BankDecision]],
    summary="决策日志",
)
async def list_decisions(bank_id: str, _user: CurrentUser):
    """列出银行决策日志 (AI 建议 vs 银行最终决策)."""
    items = await _svc().list_decisions(bank_id)
    return make_ok(items)


# ============================================================================
# 7. 提交决策
# ============================================================================

@bank_router.post(
    "/{bank_id}/decisions",
    response_model=ApiResult[BankDecision],
    status_code=status.HTTP_201_CREATED,
    summary="提交决策",
)
async def submit_decision(
    bank_id: str, payload: BankDecisionCreate, _user: CurrentUser,
):
    """提交银行决策 (根据当前信任阶段自动判定 autoHandled)."""
    decision = await _svc().submit_decision(
        bank_id=bank_id,
        enterprise_id=payload.enterprise_id,
        amount=payload.amount,
        ai_recommendation=payload.ai_recommendation,
        bank_final_decision=payload.bank_final_decision,
        enterprise_name=payload.enterprise_name,
        reason=payload.reason,
    )
    return make_ok(decision)


# ============================================================================
# 8. 统计数据
# ============================================================================

@bank_router.get(
    "/{bank_id}/statistics",
    response_model=ApiResult[BankStatistics],
    summary="统计数据",
)
async def get_statistics(bank_id: str, _user: CurrentUser):
    """银行统计数据 (总笔数/自动通过/人工通过/拒绝/总额)."""
    stats = await _svc().get_statistics(bank_id)
    return make_ok(stats)


# ============================================================================
# 9. 调整授信乘数 (APP-01 端点 9, C档独立兜底)
# ============================================================================

@bank_router.post(
    "/{bank_id}/credit-multiplier",
    response_model=ApiResult[CreditMultiplierResult],
    summary="调整授信乘数 (APP-01)",
)
async def adjust_credit_multiplier(
    bank_id: str, payload: CreditMultiplierPayload, _user: CurrentUser,
):
    """调整银行授信乘数 (C档兜底: 内存存储, 重启丢失).

    入参: multiplier (0.1-5.0), reason (审计用).
    返回: 新/旧授信额度 (分) + 应用后乘数 + 操作时间.
    """
    result = await _svc().adjust_credit_multiplier(
        bank_id=bank_id,
        multiplier=payload.multiplier,
        reason=payload.reason,
    )
    return make_ok(result)


# ============================================================================
# 10. 冻结/解冻监管账户 (APP-01 端点 10, C档独立兜底)
# ============================================================================

@bank_router.post(
    "/{bank_id}/freeze-account",
    response_model=ApiResult[FreezeAccountResult],
    summary="冻结/解冻监管账户 (APP-01)",
)
async def freeze_account(
    bank_id: str, payload: FreezeAccountPayload, _user: CurrentUser,
):
    """冻结/解冻监管账户 (C档兜底: 内存存储).

    入参 body: accountId (目标账户), freeze (true=冻结/false=解冻), reason (审计).
    账户不存在时返回 404.
    """
    try:
        result = await _svc().freeze_account(
            bank_id=bank_id,
            account_id=payload.account_id,
            freeze=payload.freeze,
            reason=payload.reason,
        )
    except KeyError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from None
    return make_ok(result)


# ============================================================================
# 11. 监管账户列表 (APP-01 端点 11, C档独立兜底)
# ============================================================================

@bank_router.get(
    "/{bank_id}/supervision-accounts",
    response_model=ApiResult[list[SupervisionAccount]],
    summary="监管账户列表 (APP-01)",
)
async def list_supervision_accounts(bank_id: str, _user: CurrentUser):
    """列出银行监管账户 (C档兜底: 内存存储).

    返回: SupervisionAccount[] (脱敏账户号/状态/余额/关联企业).
    未注册银行返回空数组 (不报错, 让前端展示空态).
    """
    items = await _svc().list_supervision_accounts(bank_id)
    return make_ok(items)
